### PR TITLE
Implement codefix for MSTEST0037: Use proper 'Assert' methods

### LINKED ISSUE #3315: new Analyzer+CodeFix: reccomend against wrong assertion method (e.g. `Assert.IsTrue(x != null)`)
## Summary
As some assertions may be written in different ways, some of them provide a better experience: better/richer message, clearer intent, better readability/maintainability.

## Background and Motivation
- default Exception-Messages are better
```cs
string text = "Mouse";
Assert.IsTrue("House" == text);
//  Assert.IsTrue failed.
Assert.AreEqual("House", text);
// Assert.AreEqual failed. Expected:<House>. Actual:<Mouse>. 
```
This is also useful when a test is failing in a pipeline, or you get the stack trace only. This way you *may* not have to dig too deep and "waste" time debugging, just to get to that point to know what `text` actually is. Or when you get a notification from the pipeline a la `test method "VeryImportantTest" failed with message "Assert.IsTrue failed."`

- we have a company-internal analyzer that does this.
- xUnit has a similar analyzers: 
  - https://xunit.net/xunit.analyzers/rules/xUnit2003
  - https://xunit.net/xunit.analyzers/rules/xUnit2004
- (though for some tests, it is explicitly wanted due to explicit testing of operator overload; but for those case the warning is just disabled via `#pragma` or Attribute)

## Proposed Feature
essentially all permutations of IsFalse/IsTrue with `is`/`is not`/`==`/`!=`
- Assert.IsNotNull, e.g.
  -  `Assert.IsTrue(x != null)` => `Assert.IsNotNull(x)`
  -  `Assert.IsTrue(null != x)` => `Assert.IsNotNull(x)`
  -  `Assert.IsFalse(x != null)` => `Assert.IsNull(x)`
  -  `Assert.IsTrue(x is not null)` => `Assert.IsNotNull(x)`
  - ...
  -  `Assert.AreNotEqual(null, x)` => `Assert.IsNotNull(x)`
  - ...
  -  either side may be `null` literal

- Assert.Are[Not]Equal, e.g.
  - `Assert.IsTrue(x == y)` => `Assert.AreEqual(x, y)`
  - `Assert.IsTrue(x != y)` => `Assert.AreNotEqual(x, y)`
  -  `Assert.IsFalse(x == y)` => `Assert.AreNotEqual(x, y)`
  -  `Assert.IsFalse(x != y)` => `Assert.AreEqual(x, y)`
  - ...
  - maybe don't be overly smart, and just let the 'const in `expected`' be handled by the other analyzer

-  Assert.IsTrue/IsFalse, e.g.
   - `Assert.AreEqual(true, b)` => `Assert.IsTrue(b)`
   - only when there is a bool const as one param
   - may be a cascaded result result from `Assert.IsTrue(true == b)`, stupid, but possible; again, just let it do one step at a time, for simplicity?

also wouldn't matter to me if they are all the same DiagnosticID, since those cases are all "wrong similarly"
## Alternative Designs

As mentioned, our analyzer grew over time and is quite complex. And I'm not sure if we have a bug or a missing case. Having such an Analyzer+CodeFix in MsTest directly would also provide a wider audience for finding such "bugs".

### PR DESCRIPTION
**Pull request by Youssef1313**
Fixes #3315

### CODE REVIEW

File: test/UnitTests/MSTest.Analyzers.UnitTests/UseProperAssertMethodsAnalyzerTests.cs [hunk 1]
```diff
@@ -354,10 +582,28 @@ public void MyTestMethod()
             }
             """;
 
-        await VerifyCS.VerifyAnalyzerAsync(
+        // error CS1503: Argument 1: cannot convert from 'object' to 'bool'
```
Comment (Evangelink) [inline test/UnitTests/MSTest.Analyzers.UnitTests/UseProperAssertMethodsAnalyzerTests.cs]: Shall we instead not raise? We could detect if that can be converted to the other type. I guess

File: test/UnitTests/MSTest.Analyzers.UnitTests/UseProperAssertMethodsAnalyzerTests.cs [hunk 2]
```diff
@@ -397,10 +646,28 @@ public void MyTestMethod()
...
```
Comment (Evangelink) [inline test/UnitTests/MSTest.Analyzers.UnitTests/UseProperAssertMethodsAnalyzerTests.cs]: Same remark about avoiding the analyzer diagnostic instead.

**Review by Evangelink — COMMENTED** [review]
For the cases leading to compilation error, could we tweak the rule not to report instead?

File: test/UnitTests/MSTest.Analyzers.UnitTests/UseProperAssertMethodsAnalyzerTests.cs (hunk 1 above)
Comment (Youssef1313) [reply test/UnitTests/MSTest.Analyzers.UnitTests/UseProperAssertMethodsAnalyzerTests.cs]: The diagnostic is still correct IMO. It's that the codefix should be producing a cast.

Comment (Youssef1313) [reply test/UnitTests/MSTest.Analyzers.UnitTests/UseProperAssertMethodsAnalyzerTests.cs]: I adjusted the codefix to produce a cast.

File: src/Analyzers/MSTest.Analyzers.CodeFixes/UseProperAssertMethodsFixer.cs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+            context.RegisterCodeFix(
+            CodeAction.Create(
+                title: string.Format(CultureInfo.InvariantCulture, CodeFixResources.UseProperAssertMethodsFix, properAssertMethodName),
+                createChangedDocument,
+                equivalenceKey: nameof(UseProperAssertMethodsFixer)),
+            diagnostic);
+        }
+    }
+
+    private static async Task<Document> FixAssertMethodForSimpleModeAsync(Document document, Location location1, Location location2, SyntaxNode root, SimpleNameSyntax simpleNameSyntax, string properAssertMethodName, CancellationToken cancellationToken)
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers.CodeFixes/UseProperAssertMethodsFixer.cs]: Could you please replace `location1` and `location2` by more meaningful names?

File: src/Analyzers/MSTest.Analyzers.CodeFixes/UseProperAssertMethodsFixer.cs [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        }
+
+        DocumentEditor editor = await DocumentEditor.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+        FixInvocationMethodName(editor, simpleNameSyntax, properAssertMethodName);
+        editor.ReplaceNode(node1, SyntaxFactory.Argument(node2).WithAdditionalAnnotations(Formatter.Annotation));
+
+        return editor.GetChangedDocument();
+    }
+
+    private static async Task<Document> FixAssertMethodForAddArgumentModeAsync(Document document, Location location1, Location location2, Location location3, SyntaxNode root, SimpleNameSyntax simpleNameSyntax, string properAssertMethodName, CancellationToken cancellationToken)
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers.CodeFixes/UseProperAssertMethodsFixer.cs]: Same remark about location variables.

File: src/Analyzers/MSTest.Analyzers.CodeFixes/UseProperAssertMethodsFixer.cs [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        SimpleNameSyntax simpleNameSyntax,
+        string properAssertMethodName,
+        bool needsNullableBoolCast,
+        CancellationToken cancellationToken)
+    {
+        // This doesn't properly handle cases like Assert.AreEqual(message: "My message", expected: true, actual: x)
+        // The proper handling of this may be Assert.IsTrue(message: "My message", condition: x)
+        // Or: Assert.IsTrue(x, "My message")
+        // For now this is not handled.
+        if (root.FindNode(additionalLocations[0].SourceSpan) is not ArgumentSyntax node1)
```
Comment (Evangelink) [inline src/Analyzers/MSTest.Analyzers.CodeFixes/UseProperAssertMethodsFixer.cs]: Same remark for `node1`

**Review by Evangelink — COMMENTED** [review]
LGTM - please rename the few variables with numbering suffix.

**Review by Evangelink — APPROVED** [review]

**Pull request merged**